synapse/storage/databases/main/deviceinbox.py

```diff
@@ -733,74 +680,94 @@ async def _remove_hidden_devices_from_device_inbox(
-                LIMIT ?
-            """
+    def _remove_devices_from_device_inbox_txn(
+        self,
+        txn: LoggingTransaction,
+        update_name: str,
+        progress: JsonDict,
+        batch_size: int,
+    ) -> Tuple[int, bool]:
+        """Remove devices that were either deleted or hidden from the device_inbox table.
 
-            txn.execute(sql, (last_stream_id, True, batch_size))
-            rows = txn.fetchall()
+        Args:
+            update_name: The name of the update to run, used to determine how to read the
+                devices table.
+            progress: The update's progress dict.
+            batch_size: The batch size for this update.
 
-            num_deleted = 0
-            for row in rows:
-                num_deleted += self.db_pool.simple_delete_txn(
-                    txn,
-                    "device_inbox",
-                    {"device_id": row[0], "user_id": row[1], "stream_id": row[2]},
-                )
+        Returns:
+            The number of rows deleted, and whether the update should be ended.
+        """
 
-            if rows:
-                # We don't just save the `stream_id` in progress as
-                # otherwise it can happen in large deployments that
-                # no change of status is visible in the log file, as
-                # it may be that the stream_id does not change in several runs
-                self.db_pool.updates._background_update_progress_txn(
-                    txn,
-                    self.REMOVE_HIDDEN_DEVICES,
-                    {
-                        "device_id": rows[-1][0],
-                        "user_id": rows[-1][1],
-                        "stream_id": rows[-1][2],
-                    },
+        if "max_stream_id" in progress:
+            max_stream_id = progress["max_stream_id"]
+        else:
+            txn.execute("SELECT max(stream_id) FROM device_inbox")
+            # There's a type mismatch here between how we want to type the row and
+            # what fetchone says it returns, but we silence it because we know that
+            # res can't be None.
+            res: Tuple[Optional[int]] = txn.fetchone()  # type: ignore[assignment]
+            if res[0] is None:
+                return 0, True
+            else:
+                max_stream_id = res[0]
+
+        start = progress.get("stream_id", 0)
+        stop = start + batch_size
+
+        nested_select = "SELECT device_id, user_id FROM devices"
+        args = (start, stop)
+        if update_name == self.REMOVE_HIDDEN_DEVICES:
+            # If we want to remove hidden devices, select only rows from the devices table
+            # that have `hidden = TRUE`.
+            nested_select += " WHERE hidden = ?"
+            # We need to ignore mypy's error here, otherwise we can't use different
+            # arguments depending on whether we want to filter for hidden devices.
+            args += (True,)  # type: ignore[assignment]
```

Comment:
think this works fine:

```suggestion
            nested_select += " WHERE hidden"
```

Reply:
Oh TIL